packages/apollo-server-lambda/src/ApolloServer.ts

```diff
@@ -29,7 +30,7 @@ export class ApolloServer extends ApolloServerBase {
   // another place, since the documentation becomes much more complicated when
   // the constructor is not longer shared between all integration
   constructor(options: Config) {
-    if (process.env.ENGINE_API_KEY || options.engine) {
+    if (process.env[keyEnvVar] || process.env[legacyKeyEnvVar] || options.engine) {
```

Comment:
@zionts pointed out that we're doing this in a number of places. I have a suggestion. Let's export a function from `apollo-server-env`, something like `getApiKeyFromEnv()`.

This fn can be responsible for:
* Throwing when both exist
* Notifying (once) that the user is depending on deprecated functionality
* Caching values requested from `process.env` and returning them. Accessing these values isn't free, and we only need to do it once in a process's lifetime.
* Returning a single apiKey, cleaning up much of the `||` logic we're seeing everywhere.

Reply:
All right, so I followed a similar pattern to what I did here for `engine.graphVariant`, which is to throw it in an exported function from `agent.ts` in apollo-engine-reporting, which gateway relies upon. The challenge here is that it's important to have `EngineReportingAgent` be a class that can run on its own and perform its own validation, which means that we don't want to just calculate the "right" thing up above and pass it through. Additionally, the gateway and aer actually have different objects that they use in their construction, so there needs to be _some_ calculation and conversion in order to satisfy the gateway.

All to say, I think having this be an exported function from aer so that aer can continue to remain a standalone package has benefits, and we can still export it and rely on it from other places, including in constructing the gateway object. That's what I've done here in response to this comment. Now we _do_ call it twice in the case of a gateway, which would, theoretically, be a fairly simple fix, but I didn't prioritize doing so because of the added complexity that it causes and the unclear impact of the benefit (one deprecation message in your console instead of two and one more read of an environment variable at boot-up time)